### PR TITLE
Atomicity of Replicated table removal.

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Bug Fix

Short description (up to few sentences):
Atomicity in drop replicated table.

Detailed description (optional):
Previously, when we tried to drop replicated table we removed its metadata locally and tried to remove metadata in Zookeeper. If Zookeeper was off, then the drop query wasn't executed, but table came to non-consistent state: it had data, but had no metadata. Now we rename table's metadata, then try to remove data (actually try to remove metadata in zookeeper). If an error occured, we rename metadata back. As the result query doesn't execute and table doesn't drop. 
If ClickHouse falls with an error during a drop query been executed, then after restart we check renamed metadata file. If it has data, then we rename metadata file or delete it otherwise. 

#6045

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Ok.

> If an error occured, we rename metadata back.

And what if we have partially removed some ZK data before the error happened?

**Comment by nikitamikhaylov** [comment]
Undefined behaviour.

**Comment by alexey-milovidov** [comment]
Ok. At least this implementation will lower probability of unrecoverable state.

**Comment by akuzm** [comment]
From the first principles, the deletion should go as follows:
1) atomically mark the table as deleted in a single source of truth,  Zookeeper in this case,
2) report that the deletion query has succeeded,
3) do whatever local garbage collection is needed, in a way resilient to restarts etc.
This is an often-used sequence for making complex data modification in a one-way, atomic fashion, that is relatively easy to reason about.

This PR seems to be doing something different, which may fix one failure but introduce another. For example, what happens when we modify metadata locally, fail to communicate with zookeeper, and the server goes down before it can change the local metadata back? In general, the human mind is too weak to contain all these intermediate states, so it's advisable not to have them at all.

File: dbms/src/Databases/DatabaseOrdinary.cpp [hunk 1]
```diff
@@ -135,7 +135,25 @@ void DatabaseOrdinary::loadTables(
         if (endsWith(dir_it.name(), ".sql.bak"))
             continue;
 
-        /// There are files .sql.tmp - delete.
+        // There are files that we tried to delete previously
+        const std::string tmp_drop_ext = ".sql.tmp_drop";
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabaseOrdinary.cpp]: Temporary object created and destroyed inside a loop. Will fix.

File: dbms/src/Databases/DatabaseOrdinary.cpp [hunk 2]
```diff
@@ -135,7 +135,25 @@ void DatabaseOrdinary::loadTables(
...
+        if (endsWith(dir_it.name(), ".sql.tmp_drop"))
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabaseOrdinary.cpp]: And it's unused.

File: dbms/src/Databases/DatabaseOrdinary.cpp [hunk 3]
```diff
@@ -135,7 +135,25 @@ void DatabaseOrdinary::loadTables(
...
+        {
+            const std::string table_name = dir_it.name().substr(0, dir_it.name().size() - tmp_drop_ext.size());
+            if (Poco::File(data_path + '/' + table_name).exists())
+            {
+                Poco::File(dir_it->path()).renameTo(table_name + ".sql");
+                LOG_WARNING(log, "Table was not dropped previously");
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabaseOrdinary.cpp]: Better to say exactly - what table.

File: dbms/src/Databases/DatabaseOrdinary.cpp [hunk 4]
```diff
@@ -302,6 +320,12 @@ void DatabaseOrdinary::removeTable(
     }
     catch (...)
     {
+        try
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabaseOrdinary.cpp]: We must log every exception.

File: dbms/src/Databases/DatabaseOrdinary.cpp [hunk 5]
```diff
@@ -302,6 +320,12 @@ void DatabaseOrdinary::removeTable(
...
+        {
+            Poco::File(table_metadata_path + ".tmp_drop").remove();
+            return;
+        }
+        catch (...) {}
```
Comment (alexey-milovidov) [inline dbms/src/Databases/DatabaseOrdinary.cpp]: We must log every exception.

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/src/Interpreters/InterpreterDropQuery.cpp [hunk 6]
```diff
@@ -90,11 +90,28 @@ BlockIO InterpreterDropQuery::executeToTable(String & database_name_, String & t
             /// If table was already dropped by anyone, an exception will be thrown
 
             auto table_lock = database_and_table.second->lockExclusively(context.getCurrentQueryId());
-            /// Delete table metadata and table itself from memory
 
+            const auto prev_metadata_name = database_and_table.first->getMetadataPath() + escapeForFileName(database_and_table.second->getTableName()) + ".sql";
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterDropQuery.cpp]: Bad code.

Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterDropQuery.cpp]: It's totally and absolutely wrong that metadata is manipulated here (it's internal detail of DatabaseOrdinary).

**Pull request merged**